src/db/file.rs

```diff
@@ -138,102 +141,96 @@ pub fn add_path_into_database<P: AsRef<Path>>(conn: &Connection,
-                    }
-                }
             } else {
-                Some(content.clone().into())
-            };
-
-            file_list_with_mimes.push((mime.clone(), file_path.clone()));
-
-            (
-                bucket_path,
-                content,
-                mime,
-            )
+                mime.to_owned()
+            }
         };
 
-        // If AWS credentials are configured, don't insert/update the database
-        if client.is_none() {
+        if let Some(client) = s3_client() {
+            let bucket_path = bucket_path.clone();
+            let content = content.clone();
+            let mime = mime.clone();
+
+            futures.push(client.put_object(PutObjectRequest {
+                bucket: "rust-docs-rs".into(),
+                key: bucket_path.clone(),
+                body: Some(content.clone().into()),
+                content_type: Some(mime.clone()),
+                ..Default::default()
+            }).map_err(move |e| {
+                log::error!("failed to upload to {}: {:?}", bucket_path, e);
+                client.put_object(PutObjectRequest {
+                    bucket: "rust-docs-rs".into(),
+                    key: bucket_path,
+                    body: Some(content.into()),
+                    content_type: Some(mime),
+                    ..Default::default()
+                }).map_err(|_| {
+                    panic!("failed to upload twice, exiting");
+                })
+            }));
+
+            if futures.len() > MAX_CONCURRENT_UPLOADS {
+                if rt.block_on(::futures::future::join_all(futures)).is_err() {
+                    panic!("upload to s3 failed");
+                }
+                futures = Vec::new();
+            }
+        } else {
+            // If AWS credentials are configured, don't insert/update the database
             // check if file already exists in database
-            let rows = try!(conn.query("SELECT COUNT(*) FROM files WHERE path = $1", &[&path]));
-
-            let content = content.expect("content never None if client is None");
+            let rows = try!(conn.query("SELECT COUNT(*) FROM files WHERE path = $1", &[&bucket_path]));
 
             if rows.get(0).get::<usize, i64>(0) == 0 {
                 try!(trans.query("INSERT INTO files (path, mime, content) VALUES ($1, $2, $3)",
-                                &[&path, &mime, &content]));
+                                &[&bucket_path, &mime, &content]));
             } else {
                 try!(trans.query("UPDATE files SET mime = $2, content = $3, date_updated = NOW() \
                                 WHERE path = $1",
-                                &[&path, &mime, &content]));
+                                &[&bucket_path, &mime, &content]));
             }
         }
+
+        file_list_with_mimes.push((mime.clone(), file_path.clone()));
+    }
+
+    if rt.block_on(::futures::future::join_all(futures)).is_err() {
```

Comment:
Part of me is wary about calling `join_all` with an empty Vec (in the case of not having AWS credentials configured). I figure it's harmless, but should we consider wrapping this block with an `if !futures.is_empty()` block?

Reply:
yeah it looks like [`block_on`](https://docs.rs/tokio/0.1.22/src/tokio/runtime/threadpool/mod.rs.html#260-270) will unnecessarily create a channel, which might be good to prevent?

Reply:
This I think would be good to implement.